daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/Evaluatable.scala

```diff
@@ -285,7 +285,17 @@ abstract class Evaluatable[+T <: AnyRef](protected val ci: DPathCompileInfo, qNa
   /**
    * Preferred for use in the runtime.
    */
-  @inline final def maybeConstant = constValue_.asInstanceOf[Maybe[T]]
+  @inline final def maybeConstant = {
+    // We could just do constValue_.asInstanceOf[Maybe[T]], but the compiled
+    // bytecode ends up allocating a Maybe. Since constant values are accessed
+    // very often, we really want to avoid that overhead. The following gets
+    // rid of those unnecessary allocations.
+    if (constValue_.isDefined) {
+      Maybe[T](constValue_.get.asInstanceOf[T])
```

Comment:
This is interesting. So yeah, if you return constValue_ and cast it, that's a generic operation, which requires a boxed object. 

We should consider if this can be caught somehow. At minimum we can search for "asInstanceOf[Maybe" and see if any others come up.

Reply:
When I was working on the DataValues class, I got into a workflow, where I would look dissassembly all the *.class files and look for allocations of DataValue. I also did a similar trick when I was looking for scala.lang.BigInt/BigDecimal.

It is probably more work then it is worth to do this in an automated fashion (particularly because removing all boxing would likely be a lot more effort then it is worth), however, it might be worth doing this as a one off. Possibly also document the process on the wiki and repeat it manually on occasion.

Reply:
There are no other places we do ``asInstanceOf[Maybe[...]]``. I'll try disassemble things and see if there's anything that jumps out.

Reply:
Disably turned up about 200 places where we allocate a Maybe. From the looks of it, it looks like it's in places where boxing is required, or might be difficult to remove. I'd reccommend we only work to remove these when they show up in profiling.

Reply:
I'm not exactly sure how to do this, so adding a note to this effect to the wiki page about performance coding is worth doing in a section (probably already is one) about avoiding excess boxing and unnecessary allocations.

Reply:
Will do. For reference, this was the command:
```bash
find daffodil.git -name '*.class' -exec javap -p -c '{}' \; > dissasembled.txt
```
From there, it's just a matter of grep, e.g.
```bash
grep -a new disassembled.txt | grep Maybe
```
Note that the -a flag is needed since there's UTF-8 characters or something that makes grep think the file is binary. Javap is really useful in figuring out why there are allocations since Scala does so much stuff behind your back.